### ISSUE TITLE
t0co doesn't update the pkl file

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#508]
In GitLab by @mcastril on Mar 17, 2020, 10:45

Hi guys

The t0co experiment is running, here you have the logs:

```
mcastril@ bscearth396:/esarchive/autosubmit/t0co/tmp/ASLOGS> ls -lrt
total 4099
-rwxrwxr-x+ 1 mcastril Earth    2220 mar 13 19:37 20200313_193734_monitor.log
-rwxrwxr-x+ 1 mcastril Earth    2218 mar 13 19:43 20200313_194336_monitor.log
-rwxrwxr-x+ 1 pechevar Earth  371570 mar 16 16:55 20200316_161729_run.log
-rwxrwxr-x+ 1 pechevar Earth    3054 mar 16 16:58 20200316_165833_set_status.log
-rwxrwxr-x+ 1 pechevar Earth    3383 mar 16 16:59 20200316_165946_set_status.log
-rwxrwxr-x+ 1 pechevar Earth  523514 mar 16 17:36 20200316_165953_run.log
-rwxr-x---+ 1 pechevar Earth       0 mar 16 22:51 submit_marenostrum4.sh
-rwxrwxr-x+ 1 pechevar Earth 2798246 mar 16 22:52 20200316_173631_run.log
-rwxrwxr-x+ 1 mcastril Earth    2218 mar 17 10:41 20200317_104154_monitor.log
```

But both the GUI and the monitor show and old status.

Here it is the pkl

```
mcastril@ bscearth396:/esarchive/autosubmit/t0co/pkl> ll
total 1025
-rw-rw-r--+ 1 pechevar Earth 136600 mar 17 10:44 job_list_t0co.pkl
-rw-r--r--+ 1 pechevar Earth   8192 mar 14 06:29 job_packages_t0co.db
-rw-rw-r--+ 1 pechevar Earth    315 mar 16 17:44 updated_list_t0co.txt_20200316_1744
```

Do you have an idea of what could be happening?

@dbeltrankyl @wuruchi @pablohe

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 17, 2020, 10:46

Please @pablohe open issues like this when you have these problems

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 17, 2020, 10:59

I see it crashed:

```
2020-03-16 22:51:44,463 Command sacct -n -X -j  9205103,9205104, -o jobid,State in mn1.bsc.es successful with out message: 9205103          FAILED
9205104          FAILED

2020-03-16 22:51:44,463 Output 9205103          FAILED
9205104          FAILED

2020-03-16 22:51:44,463 Successful check job command: sacct -n -X -j  9205103,9205104, -o jobid,State,
 output: 9205103          FAILED
9205104          FAILED

2020-03-16 22:51:44,463 Job t0co_20710101_fc0_15_POST is FAILED. Checking completed files to confirm the failure...
[WARNING] 2020-03-16 22:51:44,486 Job t0co_20710101_fc0_15_POST completion check failed. There is no COMPLETED file
2020-03-16 22:51:44,495 t0co_20710101_fc0_15_POST_STAT file have been transfered
2020-03-16 22:51:44,862 Job t0co_20710101_fc0_16_POST is FAILED. Checking completed files to confirm the failure...
[WARNING] 2020-03-16 22:51:44,865 Job t0co_20710101_fc0_16_POST completion check failed. There is no COMPLETED file
2020-03-16 22:51:44,877 t0co_20710101_fc0_16_POST_STAT file have been transfered
2020-03-16 22:51:45,396 Updating FAILED jobs
2020-03-16 22:51:45,396 Update finished
2020-03-16 22:51:45,396 Updating WAITING jobs
2020-03-16 22:51:45,397 Update finished
2020-03-16 22:51:45,418 Saving JobList: /esarchive/autosubmit/t0co/pkl/job_list_t0co.pkl
2020-03-16 22:51:45,431 Job list saved
2020-03-16 22:51:45,500
Jobs ready for marenostrum4: 0
2020-03-16 22:51:45,504 Number of jobs ready: 0
2020-03-16 22:51:45,504 Number of jobs available: 50
2020-03-16 22:51:45,508
Jobs ready for local: 0
2020-03-16 22:51:45,511 Number of jobs ready: 0
2020-03-16 22:51:45,511 Number of jobs available: 50
2020-03-16 22:51:45,513
Jobs ready for transfer_node: 0
2020-03-16 22:51:45,516 Number of jobs ready: 0
2020-03-16 22:51:45,516 Number of jobs available: 50
2020-03-16 22:51:45,519 Saving JobList: /esarchive/autosubmit/t0co/pkl/job_list_t0co.pkl
2020-03-16 22:51:45,531 Job list saved
2020-03-16 22:51:55,593 No more jobs to run.
2020-03-16 22:51:55,593 Some jobs have failed and reached maximum retrials
```

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 17, 2020, 15:08

Yes, something is wrong with that experiment.

What could we do? Maybe CREATE + RECOVERY?

**Comment by kinow** [comment]
In GitLab by @pablohe on Mar 17, 2020, 16:07

Using 3.12.1, I create and recovery (running now)

Will see.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Mar 17, 2020, 16:10

Hello @mcastril @dbeltrankyl @pablohe 

I think I found the problem:

While testing the graph generation of some experiments, I found that in some cases the dependencies form a cycle. This happens with t0co, t0cm, and some others. The problem is that `job_list.update_genealogy()` will fail in the line: `self.graph = transitive_reduction(self.graph)`; this will not stop the process, but the graph will be empty. I suppose the pkl save procedure uses `self.graph` and that might explain why it is not saving any change.

A quick fix could be to create with the no transitive flag. `-nt` perhaps? I think that should make the trick. Why is this experiment making a dependency cycle?

If you try the graph generation for http://bscesweb04.bsc.es/autosubmitapp/experiment/t0co
you will find that I have included a warning message when this happens.

**Comment by kinow** [comment]
In GitLab by @pablohe on Mar 17, 2020, 16:28

I did: 

```sh
> autosubmit setstatus t0co -fl " t0co_20710101_fc0_9_SIM t0co_20710101_fc0_10_SIM t0co_20710101_fc0_11_SIM t0co_20710101_fc0_12_SIM  t0co_20710101_fc0_13_SIM t0co_20710101_fc0_14_SIM t0co_20710101_fc0_15_SIM t0co_20710101_fc0_16_SIM t0co_20710101_fc0_17_SIM" -t COMPLETED  -s -np  
```

but when I put to run:


I get

```sh
...
650 of 705 jobs remaining (16:19)
Job t0co_20710101_fc0_9_SIM is COMPLETED
Jobs ready for marenostrum4: 3

ECT_BRANCH', 'SDEV_SPPT', 'CMOR_REALIZATION_INDEX', 'CUSTOM_DIRECTIVES', 'm', 'CMOR_ADD_STARTDATE', 'SCRATCH_DIR', 'transfer_node_RESERVATION', 'HPCPROJ', 'transfer_node_TEMP_DIR'])
[WARNING] Scripts check failed
Running after failed scripts check is at your own risk!
Autosubmit is running with v3.12.1b0


650 of 705 jobs remaining (16:19)
Job t0co_20710101_fc0_9_SIM is COMPLETED
Jobs ready for marenostrum4: 3


649 of 705 jobs remaining (16:20)
Job t0co_20710101_fc0_9_SAVEIC is QUEUING (Priority)
Job t0co_20710101_fc0_8_CMORATM is QUEUING (Priority)
Job t0co_20710101_fc0_8_CMOROCE is QUEUING (Priority)
Job t0co_20710101_fc0_10_SIM is COMPLETED
Jobs ready for marenostrum4: 3
```

and the SIM's that I was put in COMPLETED is not there.

Finally I updated using the update file in the pkl folder, it works.

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 17, 2020, 17:01

So the RECOVERY didn't put that SIMs to COMPLETED?

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 17, 2020, 17:02

Thank you @wuruchi , this is very useful information.

Where should the notificacion appear ? Or maybe it is not cyclic anymore.

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 17, 2020, 17:04

Oh, I see it now in t0cm:

```
Something has gone very wrong.

Graph generation is not possible for this experiment.
```


But the visualization was previously available in this experiment

**Comment by kinow** [comment]
In GitLab by @pablohe on Mar 17, 2020, 17:12

mentioned in issue auto-ecearth3#964

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 17, 2020, 17:12

Ok the error in t0cm is because of a new dependency introduced:

`https://earth.bsc.es/gitlab/es/auto-ecearth3/issues/964#note_73487`

But I wonder what was the problem with t0co before

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 24, 2020, 13:18

So if this is due to a cyclic dependency. @dbeltrankyl could you check if the notification for cyclic workflows has disappeared?

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 27, 2020, 18:54

@pablohe could you do a copy of t0co and put again the POST+1 dependency? I think that was the one breaking the workflow.

**Comment by kinow** [comment]
In GitLab by @pablohe on Apr 1, 2020, 15:54

Sure, I'll test it after the release is done.

**Comment by kinow** [comment]
In GitLab by @mcastril on Apr 21, 2020, 09:17

@pablohe could you test the cyclic dependency?

**Comment by kinow** [comment]
In GitLab by @pablohe on Apr 22, 2020, 10:43

Unfortunately, I removed `t0co` (it is part of HR experiment tunning).

In any case, I think it is not a problem I'll copy `t0dt` (has the same tree as `t0co`) 

The `POST+1` is in `SIM`, right?

**Comment by kinow** [comment]
In GitLab by @pablohe on Apr 22, 2020, 15:42

The POST dependency: 

```
[POST]
FILE = templates/common/common.post
DEPENDENCIES = CMOROCE CMORATM POST-1
```

Was updated as follow:

```
[POST]
FILE = templates/common/common.post
DEPENDENCIES = CMOROCE CMORATM POST+1
```

I tested with autosubmit 3.12.1:

```
bscearth315 /esarchive/autosubmit/t0dy/conf  > autosubmit create t0dy
Preparing .lock file to avoid multiple instances with same expid.

Checking configuration files...
platforms_t0dy.conf OK
jobs_t0dy.conf OK
autosubmit_t0dy.conf OK
expdef_t0dy.conf  OK
Configuration files OK

Using project folder: /esarchive/autosubmit/t0dy/proj
Loading parameters...

Creating the jobs list...
Creating jobs...
Adding dependencies...
Removing redundant dependencies...
Updating structure persistence...
[CRITICAL] An error has occurred: 
	Transitive reduction only uniquely defined on directed acyclic graphs.

```

**Comment by kinow** [comment]
In GitLab by @pablohe on Apr 24, 2020, 15:26

Ok it is fine

After I created it with the cyclic dependency 

I try to run and I get:

```
Preparing .lock file to avoid multiple instances with same experiment id


Checking configuration files...
platforms_t0dy.conf OK
jobs_t0dy.conf OK
autosubmit_t0dy.conf OK
expdef_t0dy.conf  OK
Configuration files OK

Starting job submission...
Creating jobs...
Loading JobList
Adding dependencies...
Removing redundant dependencies...
[CRITICAL] Unhandled exception on Autosubmit: Transitive reduction only uniquely defined on directed acyclic graphs.
Traceback (most recent call last):
  File "/shared/earth/software/autosubmit/3.12.0b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.0b0-py2.7.egg/autosubmit/autosubmit.py", line 417, in parse_args
    return Autosubmit.run_experiment(args.expid, args.notransitive)
  File "/shared/earth/software/autosubmit/3.12.0b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.0b0-py2.7.egg/autosubmit/autosubmit.py", line 937, in run_experiment
    job_list = Autosubmit.load_job_list(expid, as_conf, notransitive=notransitive)
  File "/shared/earth/software/autosubmit/3.12.0b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.0b0-py2.7.egg/autosubmit/autosubmit.py", line 3157, in load_job_list
    new=False, notransitive=notransitive)
  File "/shared/earth/software/autosubmit/3.12.0b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.0b0-py2.7.egg/autosubmit/job/job_list.py", line 146, in generate
    self.update_genealogy(new, notransitive)
  File "/shared/earth/software/autosubmit/3.12.0b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.0b0-py2.7.egg/autosubmit/job/job_list.py", line 901, in update_genealogy
    self.graph = transitive_reduction(self.graph)
  File "/shared/earth/software/autosubmit/3.12.0b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.0b0-py2.7.egg/autosubmit/job/job_utils.py", line 30, in transitive_reduction
    raise NetworkXError("Transitive reduction only uniquely defined on directed acyclic graphs.")
NetworkXError: Transitive reduction only uniquely defined on directed acyclic graphs.
```

**Comment by kinow** [comment]
In GitLab by @pablohe on Apr 24, 2020, 15:26

closed

**Comment by kinow** [comment]
In GitLab by @mcastril on Apr 24, 2020, 17:36

So as said in the meeting, the full test would cover also the create+change+run workflow.

**Comment by kinow** [comment]
In GitLab by @mcastril on May 5, 2020, 11:16

@pablohe did you do the last test?

**Comment by kinow** [comment]
In GitLab by @pablohe on May 5, 2020, 11:25

Yes a week ago https://earth.bsc.es/gitlab/es/autosubmit/issues/508#note_78699

**Comment by kinow** [comment]
In GitLab by @mcastril on May 5, 2020, 11:36

I read the message, but I was asking about the second test (see my previous message): https://earth.bsc.es/gitlab/es/autosubmit/issues/508#note_78753

**Comment by kinow** [comment]
In GitLab by @pablohe on May 5, 2020, 11:37

I can't create the experiment...

**Comment by kinow** [comment]
In GitLab by @mcastril on May 5, 2020, 17:54

> I don't remember, as far I remember It was running and at some point, it throws this error. 

As far as I remember the issue didn't start because of an error. We saw some strange behavior in the GUI and @wuruchi explained that the workflow had some cycles, which could not be. 

The error that you put is just starting the `run`.

> It doesn't happen again in any of my experiments (more than 20). 

It should not happen if you don't have cycles in the workflow, which we don't have now until we set correctly the dependencies.

What we were asking now was to do a test, to see if it is possible that the job may run with a cycle inside. You tested the create command with a bad configuration and it crashed on start.

The idea was to do another test with a good configuration, doing a create, starting to run, and then after the run started modify the configuration, introduce a cycle and continue to run. Then see if the error prompts or not.
